### PR TITLE
client/core: async req/ntfn handling, parallel new match negotiation

### PR DESCRIPTION
**Pull request by chappjc**
Resolves https://github.com/decred/dcrdex/issues/554 and a few untracked issues

Main changes:
- Async incoming request/notification handling in client/core, no longer blocking
- Parallel client negotiation of trades with new matches (still sequential for a given trade)
- Configurable client request expiration
- Longer request timeouts for client's init and redeem requests, where the server
  may wait up to broadcast timeout to respond.
- Longer timeout on client's coin waiters for `processAudit` and `notifyFee`
- Longer timeout on server's audit request
- Fix client outgoing messages not first being validated before writing to link

**client/core**

Incoming requests and notifications queue the call to the registered
handler function in a dedicated runner goroutine now to prevent
blocking subsequent reqs/ntfns.  If the channel buffer fills, handling
becomes blocking again, as it was before this change.

`(*dexConnection).runMatches` now starts negotiation of a trade's new
matches in parallel since `runMatches`>`tick`>`{swap,redeem,refund}Matches`
can block for a long time waiting for a response from the server.

In `handleMatchRoute`, wait briefly before `runMatches` to reduce the chance
that the server will not have yet saved the match ack sig sent just
prior.  On the server side, `processMatchAcks` now locks all of the
involved `matchTrackers` right out of the gates, before processing the
client's acks.  This is still a race, but it cannot be fully resolved
until/unless the client can get an OK/done back from it's own ack
response to the original match requests.  However, `resendPendingRequests`
will retry failed `'init'` requests so this is not a show stopper bug.

Do not try to swap/redeem/refund cancel matches in `(*trackedTrade).tick`.

Tweak `errorSet` formatting to reflect a set, and rename `ifany` to `ifAny`.

**client/{core,comms}**

Add `comms.WsConn.RequestWithTimeout`, implemented by `wsConn`.
`RequestWithTimeout` has a expire/response timeout duration and expire
callback input args.

Add `comms.DefaultRequestTimeout` for use in vanilla `Request`, and by
consumers that wish to set an expire function, but no particular timeout.

`(*dexConnection).signAndRequest` now accepts a timeout duration.

Remove `core.txWaitExpiration`, which was arbitrary and not coordinated
with the server.  Instead use broadcast timeout in:
- `finalizeSwapAction`>`signAndRequest`, where the server waits for an
  unspecified time that could be up to broadcast timeout.
- `finalizeRedeemAction`>`signAndRequest`, for the same reasons as above
- `processAudit`>`latencyQ.Wait`, since there is no reason to give up sooner
- `notifyFee`, for the same reason as `processAudit`.

However, there may be reason to timeout sooner in the init and redeem
requests if the client thinks resending the request might help.

The `DefaultRequestTimeout` is used in:
- place cancel order request
- place trade order request
- register request
- connect request (`authDEX`)
- config request (`connectDEX`)

TODO: If the server responds to init or redeem with an `Error` payload
with `Code` `TransactionUndiscovered`, the client may try to rebroadcast the
transaction that the server couldn't find, specially if the response
comes well in advance of broadcast timeout (penalization time).

**server/swap**

The audit request timeout should be up to broadcast timeout, but not for
the redemption request where the client does not need to check the coin
network.

Also hold the match mtx locked and check that the match isn't revoked
before storing results in `processInit` and `processRedeem`.

**client/comms**: marshal before attempting link write

In `client/comms.(*wsConn).Send`, marshal the `Message` first so that we
do not send junk to the peer even if it fails to marshal completely,
which `WriteJSON` does.

Test marshal error with `Send`/`Request`, and removal of response handler
in this case with `Request`.

Test `RequestWithTimeout` expiration handler and response handler removal.

### CODE REVIEW

File: client/core/core.go [hunk 1]
```diff
@@ -413,29 +423,58 @@ func (dc *dexConnection) processNomatch(oid order.OrderID) error {
...
+		errs := newErrorSet("trade %v - ", sm.tracker.ID())
+		// Begin negotiation for any msgMatches.
+		if len(sm.msgMatches) > 0 {
+			err := tracker.negotiate(sm.msgMatches)
 			if err != nil {
-				return err
-			}
-			_, err = tracker.tick()
-			if err != nil {
-				return err
```
Comment (chappjc) [inline client/core/core.go:426]: One tick failure was killing the other matches in https://github.com/decred/dcrdex/issues/554 also.

File: client/core/core.go [hunk 2]
```diff
@@ -413,29 +423,58 @@ func (dc *dexConnection) processNomatch(oid order.OrderID) error {
...
+		return errs.ifAny()
 	}
+
+	// Process the trades concurrently.
+	errChan := make(chan error)
+	for _, trade := range tradeMatches {
+		go func(trade *serverMatches) {
+			errChan <- runMatch(trade)
+		}(trade)
+	}
```
Comment (chappjc) [inline client/core/core.go:464]: `negotiate` still handles a trade's matches the same way, but the trades in `runMatches` are processed concurrently.

File: client/core/core.go [hunk 3]
```diff
@@ -1466,11 +1507,11 @@ func (c *Core) resolveActiveTrades(crypter encrypt.Crypter) (loaded int) {
...
-			locks, err := c.resumeTrades(dc, ready)
-			if err != nil {
-				details := fmt.Sprintf("Some active orders failed to resume: %v", err)
-				c.notify(newOrderNote("Order resumption error", details, db.ErrorLevel, nil))
-			}
+			locks := c.resumeTrades(dc, ready)
+			// if err != nil {
+			// 	details := fmt.Sprintf("Some active orders failed to resume: %v", err)
+			// 	c.notify(newOrderNote("Order resumption error", details, db.ErrorLevel, nil))
+			// }
```
Comment (chappjc) [inline client/core/core.go:1514]: Pretty sure this was a relic from when `resumeTrades` could actually return a non-nil error.

File: client/core/core.go [hunk 4]
```diff
@@ -1494,17 +1535,15 @@ func (c *Core) wait(assetID uint32, trigger func() (bool, error), action func(er
...
 	}
 	// Notify the server of the fee coin once there are enough confirmations.
 	req := &msgjson.NotifyFee{
 		AccountID: dc.acct.id[:],
 		CoinID:    coinID,
 	}
-	// We'll need this to validate the server's acknowledgement.
-	reqB := req.Serialize()
-	// Sign the request and send it.
-	err := stamp(dc.acct.privKey, req)
```
Comment (chappjc) [inline client/core/core.go:1507]: Was serializing the `NotifyFee` before stamping it, but the `Time` field is included in the serialization.  When the server signs the ack, the message is the *stamped* `NotifyFee`, so the signature verification in the response handler below must have been failing.

In any case, I'm not getting the "Account was registered, but DEX signature could not be verified.." warning below, so the `checkSig` is succeeding:

```
[INF] CORE: Attempting registration fee payment for SsoFtubsqCf57ZVqBkbzBSJZZ3XFnEPynWS of 100000000 units of dcr
[DBG] CORE[dcr]: Change output size = 36, addr = SseBhiCcFbQ84mTTimfMpHt2wbm5TFwdNrn
[DBG] CORE[dcr]: 3 signature cycles to converge on fees for tx 7e491d6e9b3c240b022b0b2963c58369b409270b21bb8d489a4a5c39a4c1d06a: min rate = 11, actual fee rate = 11 (2772 for 252 bytes), change = true
[INF] CORE: notify: |SUCCESS| (feepayment) Fee payment in progress - Waiting for 1 confirmations before trading at 127.0.0.1:7232
[DBG] CORE: Registration fee txn 7e491d6e9b3c240b022b0b2963c58369b409270b21bb8d489a4a5c39a4c1d06a:0 now has 1 confirmations.
[INF] CORE: Notifying dex 127.0.0.1:7232 of fee payment.
[DBG] CORE: Authenticated connection to 127.0.0.1:7232, 0 active matches
[INF] CORE: notify: |SUCCESS| (feepayment) Account registered - You may now trade at 127.0.0.1:7232
```

File: client/core/core.go [hunk 5]
```diff
@@ -1513,36 +1552,38 @@ func (c *Core) notifyFee(dc *dexConnection, coinID []byte) error {
...
-		err := resp.UnmarshalResult(ack)
-		if err != nil {
+		// Do NOT capture err in this closure.
+		if err := resp.UnmarshalResult(ack); err != nil {
 			errChan <- fmt.Errorf("notify fee result error: %v", err)
 			return
 		}
-		// If there was a serialization error, validation is skipped. A warning
-		// message was already logged.
-		if len(reqB) > 0 {
```
Comment (chappjc) [inline client/core/core.go:1526]: I'm also confused by why/how `reqB` could ever have been empty, and if it were why that would be OK.

File: client/core/core.go [hunk 6]
```diff
@@ -2113,6 +2160,8 @@ func (c *Core) dbTrackers(dc *dexConnection) (map[order.OrderID]*trackedTrade, e
 		}
 	}
 
+	// NOTE: We have not loaded old/inactive matches for active dex orders.
```
Comment (chappjc) [inline client/core/core.go:2163]: Will be obsolete with https://github.com/decred/dcrdex/pull/566 merged

File: client/core/core.go [hunk 7]
```diff
@@ -2762,22 +2835,43 @@ func handleMatchRoute(c *Core, dc *dexConnection, msg *msgjson.Message) error {
 	if err != nil {
 		return fmt.Errorf("match request parsing error: %v", err)
 	}
+
+	// Acknowledgements MUST be in the same orders as the msgjson.Matches.
 	matches, acks, err := dc.parseMatches(msgMatches, true)
 	if err != nil {
+		// Even one failed match fails them all since the server requires acks
+		// for them all, and in the same order. TODO: consider lifting this
+		// requirement, which requires changes to the server's handling.
```
Comment (chappjc) [inline client/core/core.go:2844]: This would not be a difficult change on the server, but right now Swapper.Negotiate requires it.

File: client/core/core.go [hunk 8]
```diff
@@ -2762,22 +2835,43 @@ func handleMatchRoute(c *Core, dc *dexConnection, msg *msgjson.Message) error {
...
 	}
+
+	// Send the match acknowledgments.
+	// TODO: Consider a "QueueSend" or similar, but do not bail on the matches.
 	err = dc.Send(resp)
 	if err != nil {
-		return err
+		log.Errorf("Send match response: %v", err)
+		// dc.addPendingSend(resp) // e.g.
 	}
```
Comment (chappjc) [inline client/core/core.go:2858]: I didn't like that a dropped connection or send failure for any other reason was preventing the matches from being run even though the server does need the match ack, and I still don't love that we don't have an outgoing message queue to retry sending such important responses since the client has no way presently to recover from this (the server won't send a match request again just because it timed out waiting for a response).  There's lots of room for improvement in this area.

File: client/core/core.go [hunk 9]
```diff
@@ -2762,22 +2835,43 @@ func handleMatchRoute(c *Core, dc *dexConnection, msg *msgjson.Message) error {
...
+	// Begin match negotiation.
+	//
+	// TODO: We need to know that the server has first recorded the ack
+	// signature, otherwise 'init' will fail. Until this is solved with a
+	// response to the ack, sleep for a moment if we are the maker on any of the
+	// orders. However, resendPendingRequests will retry the request when the
+	// trackedTrack ticks again when the timer fires in listen.
+	for _, m := range msgMatches {
+		if m.Side == uint8(order.Maker) {
+			time.Sleep(250 * time.Millisecond)
```
Comment (chappjc) [inline client/core/core.go:2871]: Another use case for a response to a response, or a more general "dialog" or sequence of responses.  Even if the match ack was successfully sent, `runMatches` and init broadcast/request can happen immediately for a maker's match.  This is a race with the server's handling of the match ack because until it stores the ack signature it won't be ready for the init!  It's not a terrible problem though because `resendPendingRequests` will keep trying if it does fail because of this.

I'm now trying to mitigate this race by locking the matches mutex early in `(*Swapper).processMatchAcks` (https://github.com/decred/dcrdex/pull/565/files#diff-5f1d20749025f318cd1a8ba0915ebd37R2110-R2126), but it's still a race AFAICT.  The best solution would be for the client to get confirmation that the match ack was received and recorded before attempting the init.

File: client/core/core_test.go [hunk 10]
```diff
@@ -454,7 +460,7 @@ type TXCWallet struct {
 
 func newTWallet(assetID uint32) (*xcWallet, *TXCWallet) {
 	w := &TXCWallet{
-		changeCoin: &tCoin{id: []byte{0x0a, 0x0b}},
+		changeCoin: &tCoin{id: encode.RandomBytes(36)},
```
Comment (chappjc) [inline client/core/core_test.go:463]: For coinIDString and a few other non-fatal issues.

File: client/core/trade.go [hunk 11]
```diff
@@ -63,11 +59,9 @@ func (match *matchTracker) parts() (*order.UserMatch, *db.MatchMetaData, *db.Mat
 // trackedCancel is always associated with a trackedTrade.
 type trackedCancel struct {
 	order.CancelOrder
-	preImg  order.Preimage
-	matches struct {
-		maker *msgjson.Match
-		taker *msgjson.Match
-	}
```
Comment (chappjc) [inline client/core/trade.go:70]: I can revert this if it's not an improvement, but for some reason it confused me more than having the `Match`es not in a struct.

File: client/core/trade.go [hunk 12]
```diff
@@ -380,23 +393,12 @@ func (t *trackedTrade) negotiate(msgMatches []*msgjson.Match) error {
...
-		}
-	}
-	// The filled amount includes all of the trackedTrade's matches, so the
-	// filled amount must be set, not just increased.
-	trade.SetFill(filled)
+	// NOTE: If Core was restarted, previously inactive matches are no longer in
+	// the t.matches map, preventing correct fill computation by summation of
+	// all match quantities for partially filled orders with inactive matches
+	// that had partially filled the order.
+	filled := trade.AddFill(newFill)
```
Comment (chappjc) [inline client/core/trade.go:400]: More that will be obsolete with https://github.com/decred/dcrdex/pull/566 merged.

File: client/core/trade.go [hunk 13]
```diff
@@ -876,30 +884,37 @@ func (t *trackedTrade) swapMatches(matches []*matchTracker) error {
...
 
 	// Send the swap. If the swap fails, set the failErr flag for all matches.
 	// A more sophisticated solution might involve tracking the error time too
 	// and trying again in certain circumstances.
+	swaps := &asset.Swaps{
+		Inputs:     inputs,
+		Contracts:  contracts,
+		FeeRate:    highestFeeRate,
+		LockChange: t.changeLocked,
+	}
```
Comment (chappjc) [inline client/core/trade.go:917]: None of this was incorrect, it just feels more readable this way.

File: client/core/trade.go [hunk 14]
```diff
@@ -1164,11 +1197,12 @@ func (t *trackedTrade) processAudit(msgID uint64, audit *msgjson.Audit) error {
...
 	// Get the asset.AuditInfo from the ExchangeWallet. Handle network latency.
 	// The coin waiter will run once every recheckInterval until successful or
-	// until expiration after txWaitExpiration.
+	// until expiration after broadcast timeout. There is no reason to give up
+	// on the request sooner since the server will not ask again.
 	errChan := make(chan error, 1)
 	var auditInfo asset.AuditInfo
 	t.latencyQ.Wait(&wait.Waiter{
-		Expiration: time.Now().Add(txWaitExpiration),
+		Expiration: time.Now().Add(t.broadcastTimeout()),
```
Comment (buck54321) [inline client/core/trade.go:1205]: It's not clear to me that the broadcast timeout is the right value for latency expiration. They seem unrelated. Is there a good reason to use the broadcast timeout here?

Comment (chappjc) [reply client/core/trade.go:1205]: After broadcast timeout, the waiter might as well give up because the swap is going to be revoked.  Before broadcast timeout, why give up?

Comment (chappjc) [reply client/core/trade.go:1205]: I can imagine the waiter printing "tx still not found" messages periodically so the user can check out their node's connectivity, but otherwise I don't see a reason to go with a shorter expiration.

File: client/core/core.go [hunk 15]
```diff
@@ -361,14 +364,17 @@ func (dc *dexConnection) parseMatches(msgMatches []*msgjson.Match, checkSigs boo
 		if checkSigs {
 			err := dc.acct.checkSig(sigMsg, msgMatch.Sig)
 			if err != nil {
-				return nil, nil, fmt.Errorf("parseMatches: %v", err)
+				errs = append(errs, err.Error())
+				continue
```
[the remark this answers is not in the document]
Comment (chappjc) [reply client/core/core.go:368]: Each match has it's own signature, so it seems alright to accept any messages Matches that verify ok.  Although I can't really imagine how it would happen that one fails verification but the others succeed.

File: client/core/trade.go [hunk 16]
```diff
@@ -301,25 +300,28 @@ func (t *trackedTrade) readConnectMatches(msgMatches []*msgjson.Match) {
...
 	trade := t.Trade()
 	t.mtx.Lock()
 	defer t.mtx.Unlock()
+	var newFill uint64
 	for _, msgMatch := range msgMatches {
 		if len(msgMatch.MatchID) != order.MatchIDSize {
-			return fmt.Errorf("match id of incorrect length. expected %d, got %d",
+			errs.add("match id of incorrect length. expected %d, got %d",
 				order.MatchIDSize, len(msgMatch.MatchID))
+			continue
```
[the remark this answers is not in the document]
Comment (chappjc) [reply client/core/trade.go:316]: Yeah, I think that makes sense in `negotiate`.  Will change it back here.

File: client/core/core.go [hunk 17]
```diff
@@ -413,29 +423,58 @@ func (dc *dexConnection) processNomatch(oid order.OrderID) error {
...
+			errChan <- runMatch(trade)
+		}(trade)
+	}
+
+	errs := newErrorSet("runMatches - ")
+	for oid := range tradeMatches {
+		if err := <-errChan; err != nil {
+			errs.addErr(err)
+		}
+		log.Debugf("Matches for order %v done processing.", oid)
```
[the remark this answers is not in the document]
Comment (chappjc) [reply client/core/core.go:471]: ah, map range strikes again.

File: client/core/trade.go [hunk 18]
```diff
@@ -367,92 +382,98 @@ func (t *trackedTrade) negotiate(msgMatches []*msgjson.Match) error {
...
+	// trackedTrade and the DB, and update each order's status and fill amount.
+	// First record any cancel order Match and update order status.
+	if cancelMatch != nil {
+		t.cancel.matches.maker = cancelMatch // taker is stored via processCancelMatch before negotiate
+		// Set the order status for both orders.
+		t.metaData.Status = order.OrderStatusCanceled
+		t.db.UpdateOrderStatus(t.cancel.ID(), order.OrderStatusExecuted)
+	}
+	for _, match := range newTrackers {
+		t.matches[match.id] = match
```
Comment (chappjc) [inline client/core/trade.go:408]: In line with the any bad data = all bad data philosophy, the db and matches map are not updated until after each msgMatch is verified.

File: client/core/trade.go [hunk 19]
```diff
@@ -367,92 +382,98 @@ func (t *trackedTrade) negotiate(msgMatches []*msgjson.Match) error {
 			},
 		}
 
-		// First check that this isn't a match on its own cancel order. I'm not crazy
-		// about this, but I am detecting this case right now based on the Address
-		// field being an empty string.
-		isCancel := t.cancel != nil && msgMatch.Address == ""
-		if isCancel {
-			match.setStatus(order.MatchComplete)
```
Comment (chappjc) [inline client/core/trade.go:375]: This `setStatus` must have been superfluous because the `match` for a cancel Match is not stored in the `t.matches` map or stored in the DB.  Should this cancel `msgMatch` get a MetaMatch stored in the DB too like the others?  The taker side of the match isn't stored either, so I think we aren't intending to store them, but I'm not sure.

**Comment by chappjc** [comment]
I think I'm done with updates to this one.  I went back on allowing some matches to fail based on the "any bad data = all bad data" stance, but the one error in `trade.tick()` being fatal to `(*dexConnection).runMatches` still doesn't sit perfectly well with me since those will be ticked again (retried) later.  https://github.com/decred/dcrdex/pull/565/files#diff-5c4d3ee4527d6c50d1379b8edb5ef11eR456
Open to suggestions for that, and for revising the `TestTradeTracking` test with `invalidAcker` being expected to cause an error from `handleMatchRoute` (via `runMatches>tick>swapMatches>finalizeSwapAction>signAndRequest>sendRequest>Request>invalidAcker`

File: client/core/trade.go (hunk 19 above)
Comment (chappjc) [reply client/core/trade.go:375]: This is now addressed by storing both maker and taker Matches in https://github.com/decred/dcrdex/compare/2c42ccd942492b584e539076b36fc6b8c3c08699...c952eb9e8a9e0bfb9197bff9745e94b5b230884b#diff-ca928c370ec8b724badec3db0dbd7bc4R368
I think the taker's cancel Match not being stored was a bug.

File: client/core/trade.go [hunk 20]
```diff
@@ -327,132 +322,134 @@ func (t *trackedTrade) negotiate(msgMatches []*msgjson.Match) error {
...
-					Base:  t.Base(),
-					Quote: t.Quote(),
-				},
-				Match: &order.UserMatch{
-					OrderID:     oid,
-					MatchID:     mid,
-					Quantity:    msgMatch.Quantity,
-					Rate:        msgMatch.Rate,
-					Address:     msgMatch.Address,
-					Status:      order.MatchStatus(msgMatch.Status),
```
Comment (chappjc) [inline client/core/trade.go:363]: Note that this previous code was potentially setting different Status values between the MatchMetaData and UserMatch.  Since I `negotiate` is intended to start new matches, I believe, I've updated this with a `SetStatus(order.NewlyMatched)` several lines down, while the new `makeMetaMatch` method ensures both Status values are consistent.

**Comment by chappjc** [comment]
Definitely not an issue with this PR, but `using fallback of 2000000` is indicating a crazy high fee.  We just merged https://github.com/decred/dcrdex/pull/533 which changed the fallback fee value, so something is funny with the fee rate now.  Hard to tell if it's an incorrect wallet setting.  I noticed you reviewed that PR prior to the commits that fixed the crazy high fallback fee I noticed.

**Comment by chappjc** [comment]
Yeah, DCR fallback fee rate bug when using `dexcctl` to setup wallets:

```
2020-07-31 23:24:28.835 [INF] CORE[btc]: Setting up new btc wallet at localhost:20556/wallet/gamma.
2020-07-31 23:24:28.835 [TRC] CORE[btc]: fallback fees set at 100 Satoshis/byte
2020-07-31 23:24:28.835 [INF] CORE: Loaded btc wallet configuration. Deposit address = msno8TvCPFRLMrSjoDXnK3xqmVaiaQWMoE
2020-07-31 23:24:28.835 [TRC] CORE[dcr]: fallback fees set at 2000000 atoms/byte
2020-07-31 23:24:28.835 [INF] CORE[dcr]: Setting up new DCR wallet at 127.0.0.1:19567 with TLS certificate "/home/jon/dextest/dcr/alpha/rpc.cert".
```

A temporary workaround is to go into the `Reconfigure Decred Wallet` dialog on the browser and submit with 0.0002, which should be already entered.  Doing this sets the fallback fee rate to 20 atoms/byte where is should be:

```
[TRC] CORE[dcr]: fallback fees set at 20 atoms/byte
```

**Comment by chappjc** [comment]
The incorrect DCR default for non-browser wallet setups should be fixed in the latest commit: https://github.com/decred/dcrdex/pull/565/commits/03b3452570024e2d26379305cd87fc6c23a1e63f

**Comment by chappjc** [comment]
@buck54321 Anything to address remaining?  I don't have better choices for the coinwaiter timeouts discussed in https://github.com/decred/dcrdex/pull/565#discussion_r462012215.

File: client/core/trade.go (hunk 19 above)
Comment (buck54321) [reply client/core/trade.go:382]: Thanks for this. This is the first step needed to fix https://github.com/decred/dcrdex/blob/be176bad73c8ffb765ecc118153bc690d448a5d9/client/core/core.go#L2303

File: client/core/trade.go [hunk 21]
```diff
@@ -334,132 +329,134 @@ func (t *trackedTrade) negotiate(msgMatches []*msgjson.Match) error {
...
+
+		log.Infof("Starting negotiation for match %v for order %v with swap fee rate = %v, quantity = %v",
+			mid, oid, match.MetaMatch.Match.FeeRateSwap, qty)
+		match.SetStatus(order.NewlyMatched) // these must be new matches
+		newTrackers = append(newTrackers, match)
+	}
+
+	// Now that each Match in msgMatches has been validated, store them in the
+	// trackedTrade and the DB, and update each order's status and fill amount.
+	// First record any cancel order Match and update order status.
```
Comment (buck54321) [inline client/core/trade.go:369]: Comment can be moved down a block.

File: client/core/trade.go [hunk 22]
```diff
@@ -334,132 +329,134 @@ func (t *trackedTrade) negotiate(msgMatches []*msgjson.Match) error {
...
+		if err != nil {
+			return fmt.Errorf("failed to update match in db: %w", err)
 		}
+		// Store a completed taker cancel match too.
+		takerCancelMeta := t.makeMetaMatch(t.cancel.matches.taker)
+		takerCancelMeta.SetStatus(order.MatchComplete)
+		err = t.db.UpdateMatch(takerCancelMeta)
+		if err != nil {
+			return fmt.Errorf("failed to update match in db: %w", err)
+		}
```
Comment (buck54321) [inline client/core/trade.go:388]: Why not handle this in `processCancelMatch`?

File: client/core/trade.go [hunk 23]
```diff
@@ -334,132 +329,134 @@ func (t *trackedTrade) negotiate(msgMatches []*msgjson.Match) error {
...
-	if t.cancel != nil && t.cancel.matches.maker != nil {
-		filled += t.cancel.matches.maker.Quantity
+	// NOTE: If Core was restarted, the loaded trackedTrades lack the cancel
+	// order Match data, preventing correct fill computation by summation of all
+	// t.matches quantities plus t.cancel.matches.maker.Quantity. For now, just
+	// increment the trade's FillAmt, which is stored and loaded with the order
+	// in the DB, by the newly-filled amount from these matches. Alternatively,
+	// a trade with known cancel could just be set to 100% filled without trying
+	// to do the accounting properly.
+	filled := trade.AddFill(newFill)
```
Comment (buck54321) [inline client/core/trade.go:405]: The comment is outdated since all match data is loaded now. I personally prefer setting the fill based on the known matches, but I know that you mentioned that you liked this incremental approach. Got a good reason? Counting / setting based on match iteration is idempotent, so perhaps less error-prone.

File: client/db/types.go [hunk 24]
```diff
@@ -193,6 +193,12 @@ type MetaMatch struct {
 	Match *order.UserMatch
 }
 
+// SetStatus sets the match status in both the UserMatch and the MatchMetaData.
+func (m *MetaMatch) SetStatus(status order.MatchStatus) {
```
Comment (buck54321) [inline client/db/types.go:197]: Why is this exported now?

File: client/core/trade.go (hunk 23 above)
Comment (chappjc) [reply client/core/trade.go:405]: Right, the data is loaded now, just missing the cancel taker which isn't part of the sum.

My feeling on the incremental fill accounting is that losing track of a match won't break the sum, but I take your point that by doing SetFill with the sum you will not risk adding fill twice (idem-potency).  I'll switch back.  It's not a big issue to me.

File: client/core/trade.go (hunk 22 above)
Comment (chappjc) [reply client/core/trade.go:388]: I wanted to update the db at roughly the same time, as the comment at the end `processCancelMatch` says without any real justification though.  `processCancelMatch` and `negotiate` are run in sequence, so I'll just do it in `processCancelMatch`.

File: client/db/types.go (hunk 24 above)
Comment (chappjc) [reply client/db/types.go:197]: Used from core/trade.go via `(*matchTracker).SetStatus`, and directly via `(*db.MetaMatch).SetStatus` for the maker/taker cancel match meta data.

Comment (buck54321) [reply client/db/types.go:197]: Derp. I was reading this wrong. This is good.

**Comment by chappjc** [comment]
@buck54321 Back to `SetFill`.  I need to manually retest with a restart with a partially filled trade though.

**Comment by chappjc** [comment]
Retested and working as expected still:

Partially filled book order;

image

Restart dexc, then fill the rest of it:

image

Fill at 100%.

Also note the one canceled order showing 100% fill as expected with the fill sum computation in `negotiate` considering the maker cancel match quantity in the fill amount.

File: client/core/core.go [hunk 25]
```diff
@@ -2916,22 +2987,47 @@ func handleMatchRoute(c *Core, dc *dexConnection, msg *msgjson.Message) error {
...
-	defer c.refreshUser()
+	defer c.refreshUser() // update WalletState
 
+	// Begin match negotiation.
+	//
+	// TODO: We need to know that the server has first recorded the ack
+	// signature, otherwise 'init' will fail. Until this is solved with a
+	// response to the ack, sleep for a moment if we are the maker on any of the
+	// orders. However, resendPendingRequests will retry the request when the
+	// trackedTrack ticks again when the timer fires in listen.
```
Comment (itswisdomagain) [inline client/core/core.go:3024]:
```suggestion
	// trackedTrade ticks again when the timer fires in listen.
```

File: client/core/core.go (hunk 8 above)
Comment (itswisdomagain) [reply client/core/core.go:3014]: Right. This is documented under **Potential causes of trade disruptions** in #361 which is a little similar to a previously existing issue where failure to send an init or redeem request to the server would cause the trade to go to hell. The latter is fixed with `resendPendingRequests` and similar stuff can be done for request acks send errors. Can create an issue to track this.

Comment (chappjc) [reply client/core/core.go:3014]: Nice, thanks @itswisdomagain

**Comment by chappjc** [comment]
@itswisdomagain Please to note any remaining issues to address in a follow up. I've retested again, and we need the DCR fee rate fix, so I'm going to get this merged now.
